supply/lib/supply/uploader.rb

```diff
@@ -81,6 +85,21 @@ def perform_upload_meta(version_codes, track_name)
           version_codes.to_s == ""
         end
 
+        upload_worker = FastlaneCore::QueueWorker.new do |job|
+          begin
+            UI.message("Preparing uploads for language '#{job.language}'...")
+            start_time = Time.now
+            listing = client.listing_for_language(job.language)
+            upload_metadata(job.language, listing) unless Supply.config[:skip_upload_metadata]
+            upload_images(job.language) unless Supply.config[:skip_upload_images]
+            upload_screenshots(job.language) unless Supply.config[:skip_upload_screenshots]
+            job.release_notes << upload_changelog(job.language, job.version_code) unless Supply.config[:skip_upload_changelogs]
+            UI.message("Uploaded all items for language '#{job.language}'... (#{Time.now - start_time} secs)")
+          rescue => error
+            UI.error("#{job.language} - #{error}")
```

Comment:
From my experience I would rather let `supply` upload all available assets and fix the missing ones later (w or w/o another fastlane call). The reason is because of the team I'm working with (_Marketing_) is prone to many human errors that invalidates the data we send to the Google Play Console (like title too long, invalid screenshot resolution, etc) and they prefer manual fixes in the UI rather in an automated system 🤷🏻‍♂️  

Also if I'm not mistaken, I think this is the default behaviour of `deliver` as well (_continue all uploads and stop with error at the end_) 🤔 

--- 

> that means your CI step will show up green making you think that everything went well, while ultimately your end users will have some screenshots or images or release notes missing.

Tested locally and the exit code is `1` when there is an upload error, so we should be covered in CI when an error is thrown. Do you mean that when setting `changes_not_sent_for_review` to `false` and the `track` is `production` the process will continue to submit the app to the store even if some uploads failed?

Reply:
I did a couple more tests and realized that any network or API errors will actually be logged and ignored during the upload process. For instance, the following error was thrown but the process exited with code 0.
```
[21:24:37]: fr-FR - Google Api Error: Unauthorized - Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
```

The errors I saw in the past tests which made me believe the process exited win non zero code were actually thrown later by `client.commit_current_edit`.

In this case, I propose to initialize a "has_errors" bool value before starting the uploads, set it to true when we catch an error, then check its value after the uploads complete and use a `UI.abort_with_message` to inform the user to check the (above) logs for any errors. How does that sound @AliSoftware?

Reply:
> For instance, the following error was thrown

I know it's not the topic of your comment, but FYI the "Google Api Error" you are seeing in your tests are likely unrelated to the changes in your PR btw, but more likely related to [a current issue apparently randomly happening for everyone with the Google Api and fastlane](https://github.com/fastlane/fastlane/issues/21507) — I myself have been having this issue all day even with fastlane `2.214.0` while it worked a couple of days ago.

> In this case, I propose to initialize a "has_errors" bool value before starting the uploads, set it to true when we catch an error, then check its value after the uploads complete and use a UI.abort_with_message to inform the user to check the (above) logs for any errors. How does that sound @AliSoftware?

Just to be clear, you'd `UI.abord_with_message!` only after having called `client.commit_current_edit` though, which means that the action would exit with non-zero code… but the changes would still have been submitted to Google Play, right?

I can see why you'd want that, but what I don't like with this approach is that it differs from how things were behaving before. In current version of _fastlane_ if one upload fails, then the action stops without trying further to upload other images nor committing the edit and submitting that passed and skipping the uploads that failed. So your suggestion would change that behavior.

---

Given that, but also given that for [the current unrelated issue we're having with Google API](https://github.com/fastlane/fastlane/issues/21507) I was also considering implementing some auto-retry mechanism, I'd then suggest to:
 - Make this current PR keep the same behavior on failure as what we currently have without parallel uploads. That is, if one upload fails, stop everything without trying to upload the rest
 - Separately, make a PR to add some `ConfigItem.new(key: :max_retries)` to the `upload_to_play_store` action to allow retrying each API calls a max number of times if one fails. Which should already reduce the risk of a whole batch of screenshot uploads failing mid-process, and would also help mitigating https://github.com/fastlane/fastlane/issues/21507
 - Separately again, make a PR to add `ConfigItem.new(key: allow_submit_partial_changes)` (happy if you have a better name idea for that one 😄 ) that would allow the behavior you're after (i.e. allowing the screenshots that passed upload to be submitted even if other screenshots failed). But that would be an opt-in option, so that everyone could choose the behavior they prefer in case of failures (you're probably turn it on while I'd personally prefer it off), and also so that we don't break backwards compatibility in the action's behavior.

Reply:
Thanks for the heads-up about the google api error, I thought it was just me 😄 

> Just to be clear, you'd UI.abord_with_message! only after having called client.commit_current_edit though, which means that the action would exit with non-zero code… but the changes would still have been submitted to Google Play, right?

Actually not, what I was thinking is adding an extra line immediately after starting the worker processes (line 115) to check for the error (something that looks like this):
```ruby
upload_worker.start
UI.abort_with_message!("Some of the upload processes ended with failure! Check the logs above for more infromation.") if has_errors
```

---

The last couple of points you made make sense to me now, so in this PR I'll focus to keep the current behaviour then do a follow up in another PR with the "allow submit partial changes" functionality.